core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandler.java

```diff
@@ -284,6 +296,103 @@ private Token getRoutingToken(Statement statement) {
     return tokenMap == null ? null : ((DefaultTokenMap) tokenMap).getTokenFactory().hash(key);
   }
 
+  private CqlIdentifier getTabletRoutingKeyspace(Statement statement) {
+    if (statement == null) {
+      if (initialStatement == null) {
+        return null;
+      }
+      statement = initialStatement;
+    }
+    ColumnDefinitions cdefs = null;
+    CqlIdentifier result = null;
+    if (statement instanceof BoundStatement) {
+      cdefs = ((BoundStatement) statement).getPreparedStatement().getVariableDefinitions();
+    } else if (statement instanceof PreparedStatement) {
+      cdefs = ((PreparedStatement) statement).getVariableDefinitions();
+    }
+    if (cdefs != null && cdefs.size() > 0) {
+      result = cdefs.get(0).getKeyspace();
+    }
+    if (result == null) {
+      return keyspace;
+    } else {
+      return result;
+    }
+  }
+
+  private CqlIdentifier getTabletRoutingTable(Statement statement) {
+    if (statement == null) {
+      if (initialStatement == null) {
+        return null;
+      }
+      statement = initialStatement;
+    }
+    if (statement instanceof BoundStatement) {
+      return ((BoundStatement) statement).getPreparedStatement().getTable();
+    } else if (statement instanceof PreparedStatement) {
+      return ((PreparedStatement) statement).getTable();
+    } else {
+      return null;
+    }
+  }
+
+  public Integer getShardFromTabletMap(Statement statement, Node node, Token token) {
+    TabletMap tabletMap = context.getMetadataManager().getMetadata().getTabletMap();
+    if (!(token instanceof TokenLong64)) {
+      LOG.trace(
+          "Token ({}) is not a TokenLong64. Not performing tablet shard lookup for statement {}.",
+          token,
+          statement);
+      return null;
+    }
+    if (tabletMap == null) {
+      LOG.trace(
+          "Could not determine shard for token {} on host {} because tablets metadata is currently null. "
+              + "Returning null.",
+          token,
+          node);
+      return null;
+    }
+    if ((getTabletRoutingKeyspace(statement) == null && keyspace == null)
+        || getTabletRoutingTable(statement) == null) {
+      return null;
+    }
+    UUID targetHostUuid = node.getHostId();
+    long tokenValue = ((TokenLong64) token).getValue();
+    String statementKeyspace =
+        getTabletRoutingKeyspace(statement) != null
+            ? getTabletRoutingKeyspace(statement).asInternal()
+            : keyspace.asInternal();
```

Comment:
Why do we need to check `keyspace` here manually? Can't `getTabletRoutingKeyspace` handle it?

Reply:
In a newer version it now does handle it